### ISSUE TITLE
Schema Migration Overriding  exist Indexes

### ISSUE THREAD

**Issue by badboy-tian** [parse-community/parse-server#9112]
### New Issue Checklist

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Before posting search [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Schema Migration Overriding  exist Indexes
### Steps to reproduce

1.create some index in mongodb tools
![image](https://github.com/parse-community/parse-server/assets/2835129/3d4a151f-911b-426d-ae67-baffcbae9738)
2.define ExportTool class in schema 
```
export const ExportTotal = {
    className: "ExportTotal",
    fields: {
        myid: {type: "Number", defaultValue: 0},
        money: {type: "Number", defaultValue: 0},
        type: {type: "String"},
        user: {type: "Pointer", targetClass: "_User"},
    },
    
    classLevelPermissions: {
        find: {},
        count: {},
        get: {},
        update: {},
        create: {},
        delete: {},
        protectedFields: {},
    },
}
```
3.config the parse-server
```
schema: {
        definitions:[ExportTool],
        lockSchemas: false,
        strict: true,
        recreateModifiedFields: true,
        deleteExtraFields: true,
    },
```
![image](https://github.com/parse-community/parse-server/assets/2835129/1043ba2e-193a-4e2b-96b8-50c7ff4354f7)

### Actual Outcome

when parse server running, the exist index has dispear
### Expected Outcome

when parse server running, the exist index shoud not dispear
### Environment

Server
- Parse Server version: `7.0.0`
- Operating system: `macos`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `local`

Database
- System (MongoDB or Postgres): `mongodb`
- Database version: `7.0.8`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `local`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `no`
- SDK version: `no`

### Logs

```
info: 11:34:54 -> Node app is running on port: 1337
[TS] info: Running Migrations
[TS] info: Running Migrations Completed

```

**Comment by mtrezza** [comment]
What is the expected behavior according to the defined schema docs? I couldn't any specific mention of this in the [docs](https://docs.parseplatform.org/defined-schema/guide/#defined-schemas-with-existing-installation).

**Comment by badboy-tian** [comment]
@mtrezza I hope the "Schema Migration" doesn't overwrite the Index I created with the mongodb tool!

**Comment by badboy-tian** [comment]
@mtrezza because the follow can not create unique index
```
indexes: {
    tagsIndex: { 
                          tags: 1 ,
                          // unique: true ?
                      } 
```
so i must create unique index in mongodb shell or mongodb tool.
then i run the "the defined schema" in parse-server, when log
```
info: 11:34:54 -> Node app is running on port: 1337
[TS] info: Running Migrations
[TS] info: Running Migrations Completed
```
the unique index created by mongodb shell was dispear.... Overriding ?

**Comment by mtrezza** [comment]
Some behavior is not explicitly described for indexes in the docs. For example if you set `strict: true`, then I would assume that indexes that aren't defined in the schema would be deleted, just like classes, fields, CPLs. In see in your case you have also set this option.

Did you take a look at the code to investigate what the current behavior is regarding indexes, so we can update the docs with the status quo? Then we can think about what we want to change.

**Comment by badboy-tian** [comment]
@mtrezza yes, when set ```strict: false``` the exist indexes was not overrided !

but because the index define in  the define schema  can not support create unique index, so we have to create unique index in mongodb tools. very inconvenient.

is there any way to supprot this?
```
indexes: {
   tag_index: {
          tag: 1, 
          unique: true, 
          background: true,
          sparse: true
    }

}
```

thank you

**Comment by rgunindi** [comment]
> Some behavior is not explicitly described for indexes in the docs. For example if you set `strict: true`, then I would assume that indexes that aren't defined in the schema would be deleted, just like classes, fields, CPLs. In see in your case you have also set this option.
> 
> Did you take a look at the code to investigate what the current behavior is regarding indexes, so we can update the docs with the status quo? Then we can think about what we want to change.

And also for strict true or false whatever result the same[also with another options]. I debugged with changing options but the result will be the same: overrides indexes. @mtrezza

**Comment by mtrezza** [comment]
> when set strict: false the exist indexes was not overrided !

> for strict true or false whatever result the same ... overrides indexes

Which one is true?

**Comment by badboy-tian** [comment]
perhaps there is a unknow bug. i set strict: false, the index also overrided , i do not know why

**Comment by badboy-tian** [comment]
https://github.com/parse-community/parse-server/assets/2835129/d5c55474-246a-4199-b72e-409aba01098a

@mtrezza @rgunindi  please see the screen record, set strict: false, also override the exist index

**Comment by badboy-tian** [comment]
@mtrezza @rgunindi

**Comment by rgunindi** [comment]
> 
> 
> https://github.com/parse-community/parse-server/assets/2835129/d5c55474-246a-4199-b72e-409aba01098a
> 
> 
> 
> @mtrezza @rgunindi  please see the screen record, set strict: false, also override the exist index
> 
> 

Yes ı said, for strict true or false whatever result the same ... overrides indexes

**Comment by badboy-tian** [comment]
@rgunindi yes, you are right

**Comment by badboy-tian** [comment]
any update ?

**Comment by rgunindi** [comment]
Don't expect early action 😁 Will take so long time

**Comment by badboy-tian** [comment]
@rgunindi yes. i have to create schema use api, create index use mongodb shell

**Comment by badboy-tian** [comment]
I find anthor way to merge schema, and the indexs can not override
@rgunindi 
example:
```
export const IMessageReply = {
    className: "IMessageReply",
    fields: {
        fromUser: {type: "Pointer", targetClass: "_User"},
        message: {type: "Pointer", targetClass: "IMessage"},
        content: {type: "String"},
    },

    classLevelPermissions: {
        find: {},
        count: {},
        get: {},
        update: {},
        create: {},
        delete: {},
        protectedFields: {},
    },
}

var schemas = [IMessageReply, ....];
```
and then
```
router.get("/schema", async (req, res) => {
    for (let i = 0; i < schemas.length; i++) {
        const className = schemas[i].className;
        const fileds = schemas[i].fields;
        const keys = Object.keys(fileds);
        console.log("className: " + className);
        const schema = new Parse.Schema(className);

        var needCreate = true;
        try {
            const oldSchema = await schema.get();
            needCreate = false;
        } catch (e) {
            needCreate = true;
        }

        for (let j = 0; j < keys.length; j++) {
            const key = keys[j];
            const value = fileds[key];
            const type = value.type;
            const defaultValue = value.defaultValue;
            const targetClass = value.targetClass;
            if (type === "Number") {
                if (defaultValue) {
                    schema.addNumber(key, {defaultValue: defaultValue});
                } else {
                    schema.addNumber(key);
                }
            } else if (type === "Boolean") {
                if (defaultValue) {
                    schema.addBoolean(key, {defaultValue: defaultValue});
                } else {
                    schema.addBoolean(key);
                }
            } else if (type === "String") {
                if (defaultValue) {
                    schema.addString(key, {defaultValue: defaultValue});
                } else {
                    schema.addString(key);
                }
            } else if (type === "Date") {
                if (defaultValue) {
                    schema.addDate(key, {defaultValue: defaultValue});
                } else {
                    schema.addDate(key);
                }
            } else if (type === "Array") {
                if (defaultValue) {
                    schema.addArray(key, {defaultValue: defaultValue});
                } else {
                    schema.addArray(key);
                }
            } else if (type === "Pointer") {
                if (defaultValue) {
                    schema.addPointer(key, targetClass, {defaultValue: defaultValue});
                } else {
                    schema.addPointer(key, targetClass);
                }
            }
        }

        if (needCreate) {
            await schema.save();
            logInfo("create " + className + " success");
        } else {
            await schema.update();
            logInfo("update " + className + " success");
        }
    }

    logInfo("create schema success");
    res.send("Hello World");
})
```
if schema exist, call schema.update, or schema.save.

**Comment by RahulLanjewar93** [comment]
https://github.com/parse-community/parse-server/blob/82fdb0d95aa78534364e95efdde39d006f883d96/src/SchemaMigrations/DefinedSchemas.js#L330-L366

this code block is responsible for deletion of indexes if they are not mentioned in the schema defined by the user.
There are cases wher parse is not able to create unique/ttl/deep indexes and the dev may have to create it themselves. 
This code block basically deletes those indexes and can create a problem 

This should ideally be behind a setting. 
Happy to create a pr if required @mtrezza

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.3.0-alpha.1](https://github.com/parse-community/parse-server/releases/tag/8.3.0-alpha.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.3.0](https://github.com/parse-community/parse-server/releases/tag/8.3.0)
